### PR TITLE
Add DevMode actions SetPlayerState and SetBallState

### PR DESCRIPTION
**Pull request by michaelbaudino**
After my first 2 attempts at trying to fast-forward a FUMBBL game up to a point of the match that would constitute an interesting-enough situation to be a chess-like puzzle, I abandoned my former ideas thanks to your advice @cmelchior (the first one was to place players/ball during the pregame sequence directly in their puzzle positions but it required to [disable setup validations](https://github.com/cmelchior/jervis-ffb/pull/46), which you ruled too reckless ; and the second was to [replay the whole FUMBBL match](https://github.com/cmelchior/jervis-ffb/pull/46) up to the puzzle point but you righteously warned that it would be a gigantic work).

Anyway, following you suggestions, I figured out how to:
1. fake as much of the pregame sequence as possible (including players/ball placement)
2. fake as much of the kickoff sequence as possible
3. then only can I finally "teleport" players/ball to their puzzle positions using the `DevMode` actions without being crippled by setup validations (3 on the LoS, everyone in their pitch half, …)

It works very well, but in order to do that, I had to enhance what the `DevMode` can do, by extending the available actions with 3 new ones (that mirror what already existing commands were doing internally), which is exactly what this PR is about.

I remember you mentioning that it was "likely that there will eventually be a `MovePlayer` dev action" in #46, so I hope what I propose here is somehow close to what you had in mind :crossed_fingers: 

Thus, this PR:
- Adds `SetPlayerLocation` DevMode action that positions a player on the pitch via the existing `SetPlayerLocation` command (and validates that the coordinate is within pitch bounds: we never know).
- Add `SetPlayerState` DevMode action that sets a player's `PlayerState` (`PRONE`, `STANDING`, `STUNNED`, …), optionally adjusting `hasTackleZones`, and clearing `intermediateState`.
- Add `SetBallLocation` DevMode action for placing the ball (with optional carrier and ball state).
- Register all three actions in `GameEngineController` dispatch.
- Add serialization support for new actions in `generatedSerializer.kt`.
- Add 11 unit tests covering state mutation, undo for each action, out-of-bounds rejection, and edge cases (`intermediateState` reset, `hasTackleZones` preservation when `null`).

> [!NOTE]
> Again, I'm usually not a Kotlin developer, so I simply tried to follow the existing conventions I could identify. Please feel free to let me know how all of this can be improved (or even feel free to push new commits to this branch if you like).

### CODE REVIEW

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt [hunk 1]
```diff
@@ -59,3 +63,38 @@ data class RemovePlayerKeyword(
 ): DevModeGameAction {
     fun getPlayer(state: Game): Player = state.getPlayerById(playerId)
 }
+
+@Serializable
+data class SetPlayerLocation(
```
Comment (cmelchior) [inline modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt]: I have been thinking a bit about how best to represent this API, and I think I came to the conclusion that exposing these low-level actions are a bit problematic as it makes it too easy to get into a wrong state. Also, this API doesn't support moving between the Dogout and the Pitch.

I think a better API for this would be something like:

```
data class SetPlayerState(
  val player: PlayerId,
  val state: PlayerState,
  val location: Location,
): DevModeGameAction
```

This will enable all the use cases in a single action + make it easier to detect invalid combinations, like setting the `PlayerState.DEAD` but not being moved to the `DogOut`.

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt [hunk 2]
```diff
@@ -59,3 +63,38 @@ data class RemovePlayerKeyword(
...
+    val playerId: PlayerId,
+    val x: Int,
+    val y: Int
+) : DevModeGameAction {
+    fun getPlayer(state: Game): Player = state.getPlayerById(playerId)
+    fun coordinate(): PitchCoordinate = PitchCoordinate(x, y)
+}
+
+@Serializable
+data class SetPlayerState(
```
Comment (cmelchior) [inline modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt:71]: This should be folded into the above described API

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt [hunk 3]
```diff
@@ -59,3 +63,38 @@ data class RemovePlayerKeyword(
...
+    val playerId: PlayerId,
+    val state: PlayerState,
+    val hasTackleZones: Boolean? = null
+) : DevModeGameAction {
+    fun getPlayer(state: Game): Player = state.getPlayerById(playerId)
+}
+
+@Serializable
+data class SetBallLocation(
```
Comment (cmelchior) [inline modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt]: Similar to the above player API, a lot of the same concernes apply here. I would rename this to:

```
data class SetBallState(
  val ball: BallId, // New ID that needs to be introduced for all balls so we can identify them across messages
  val location: OnPitchLocation,
  val ballState: BallState? = null,   
  val carriedBy: PlayerId? = null,
): DevModeGameAction

```

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/GameEngineController.kt [hunk 4]
```diff
@@ -419,6 +425,40 @@ class GameEngineController(
...
+                val coord = action.coordinate()
+                if (coord.isOutOfBounds(state.rules)) {
+                    throw InvalidActionException("Cannot place player at out-of-bounds coordinate: $coord")
+                }
+                com.jervisffb.engine.commands.SetPlayerLocation(player, coord)
+            }
+            is SetPlayerState -> {
+                val player = action.getPlayer(state)
+                val playerState = action.state
+                com.jervisffb.engine.commands.SetPlayerState(player, playerState, action.hasTackleZones)
```
Comment (cmelchior) [inline modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/GameEngineController.kt]: With the new proposed API, you can use `compositeGameAction` to bundle several commands as needed

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/GameEngineController.kt [hunk 5]
```diff
@@ -419,6 +425,40 @@ class GameEngineController(
...
+                }
+                // 2. Set ball state (ON_GROUND, CARRIED, IN_AIR)
+                val bs = action.ballState
+                val carrier = action.getPlayer(state)
+                if (bs != null) {
+                    cmd = cmd + when (bs) {
+                        BallState.CARRIED -> SetBallState.carried(ball, carrier!!)
+                        BallState.ON_GROUND -> SetBallState.onGround(ball)
+                        BallState.IN_AIR -> SetBallState.inAir(ball)
+                        else -> SetBallState.onGround(ball)
```
Comment (cmelchior) [inline modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/GameEngineController.kt]: Don't use `else` if you can avoid it. By enumerating all options, we ensure the compiler catches it if a new state is ever added, so the logic can be fixed

File: modules/jervis-engine/src/commonTest/kotlin/com/jervisffb/test/GameEngineControllerTests.kt [hunk 6]
```diff
@@ -26,13 +36,28 @@ class GameEngineControllerTests {
 
     private lateinit var controller: GameEngineController
 
-    private fun createGameController(rules: Rules): GameEngineController {
+    private fun createGameController(rules: Rules = StandardBB2020Rules().update {
+        allowPlayerEditsDuringGame = true
```
Comment (cmelchior) [inline modules/jervis-engine/src/commonTest/kotlin/com/jervisffb/test/GameEngineControllerTests.kt]: Instead of modifying the existing GameEngineControllerTests, I would add a new `DevModeGameEngineControllerTests`. This way we can have a "clean" class and one for testing the dev mode specifically.

**Review by cmelchior — CHANGES_REQUESTED** [review]
Thanks a lot for adding these 🥳 . I have some suggestions for a slightly better API, which also require some more tests to catch invalid state combinations. But the current approach are generally good.

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt (hunk 3 above)
Comment (cmelchior) [reply modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt]: `BallId` was added in https://github.com/cmelchior/jervis-ffb/commit/c348218b21a5017d43a7ca899d96ea42fd7553ea

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt (hunk 2 above)
Comment (michaelbaudino) [reply modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt:71]: Your suggested merged API above does not mention `hasTackleZone`: should we add it to you suggested API, or can we infer it from the player state (standing → tackle zone, everything else → no tackle zone)?

Also, what about states defined in `PlayerStatusEffectType`, like `ROOTED` or `DISTRACTED`? Should we be able to set them with this merged `SetPlayerState` API?

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt (hunk 1 above)
Comment (michaelbaudino) [reply modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt]: Thanks, I'll try to implement this API 👍

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt (hunk 2 above)
Comment (cmelchior) [reply modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt:71]: Hmm, good question. For now, I think we can infer the tackle zone state from the PlayerState. It is technically possible to have them change independently from each other, but I don't think that is relevant for this API. At least not right now.

With regard to Rooted or Distracted, it should probably be its own `SetPlayerStatusEffect` command that also automatically infer the tacklezone where relevant.

Comment (michaelbaudino) [reply modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt:71]: I think I addressed all your comments (except for `SetPlayerStatusEffect` that I'd like to add in a separate PR, if that's OK for you: I don't yet have a use case for it).

What do you think?

File: modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt [hunk 7]
```diff
@@ -59,3 +66,42 @@ data class RemovePlayerKeyword(
 ): DevModeGameAction {
     fun getPlayer(state: Game): Player = state.getPlayerById(playerId)
 }
+
+@Serializable
+data class SetPlayerState(
+    val playerId: PlayerId,
+    val state: PlayerState,
+    val x: Int,
+    val y: Int,
```
Comment (cmelchior) [inline modules/jervis-engine/src/commonMain/kotlin/com/jervisffb/engine/actions/DevModeGameAction.kt:75]: This API still has some issues, e.g., you need to fake these when choosing a Dogout state, but it is minor, so probably fine to merge as is

**Review by cmelchior — APPROVED** [review]
